File: mobile/lib/src/features/runtime/infra/workspace_buffer_guard_request.dart
typedef WorkspaceOperationRequest = Future<Object?> Function(
  String type,
  Map<String, Object?> payload,
  Duration? timeout,
);

Future<Object?> requestWithWorkspaceBufferGuard({
  required WorkspaceOperationRequest request,
  required String workspaceId,
  required String operation,
  required Map<String, Object?> payload,
  required Duration timeout,
}) async {
  var status = _guardResponse(
    await request('workspace.bufferGuard.acquire', {
      'id': workspaceId,
      'operation': operation,
    }, const Duration(seconds: 15)),
  );
  final id = status['guardId'];
  if (id is! String || id.isEmpty) {
    throw const FormatException('The runtime did not return a buffer guard.');
  }
  final elapsed = Stopwatch()..start();
  try {
    while (status['ready'] != true) {
      final blockers = status['blockers'];
      if (blockers is! List) {
        throw const FormatException(
          'The runtime did not verify editor buffers.',
        );
      }
      if (blockers.isNotEmpty) {
        throw StateError(
          'Resolve editor buffers on the connected clients before continuing:\n${blockers.map((blocker) {
            final entry = _guardResponse(blocker);
            return '${entry['path']}: ${entry['reason']}';
          }).join('\n')}',
        );
      }
      if (status['disconnectedClients'] != 0) {
        throw StateError(
          'A client disconnected during buffer verification. Prepare the operation again.',
        );
      }
      if (elapsed.elapsed >= const Duration(seconds: 20)) {
        throw StateError(
          'Some connected editors did not respond. The workspace was preserved.',
        );
      }
      await Future.pause(const Duration(milliseconds: 100));
      status = _guardResponse(
        await request('workspace.bufferGuard.status', {
          'guardId': id,
        }, const Duration(seconds: 5)),
      );
    }
    return await request('workspace.$operation', {
      ...payload,
      'bufferGuardId': id,
    }, timeout);
  } finally {
    elapsed.stop();
    try {
      await request('workspace.bufferGuard.release', {
        'guardId': id,
      }, const Duration(seconds: 5));
    } on Object {
      // A running mutation owns the lock until it finishes, including when
      // this caller timed out. Pending preparations also expire on the host.
    }
  }
}

Map<String, Object?> _guardResponse(Object? value) {
  if (value is! Map) {
    throw const FormatException('Invalid buffer verification response.');
  }
  return Map<String, Object?>.from(value);
}

File: rust/alera-cli/src/workspace_buffer_guard_request.rs
use anyhow::{anyhow, bail, Result};
use serde_json::{json, Value};

use crate::runtime_host_client::RuntimeHostRpcClient;

pub async fn request_with_workspace_buffer_guard(
    client: &mut RuntimeHostRpcClient,
    operation: &str,
    payload: &Value,
) -> Result<Value> {
    let workspace_id = payload
        .get("id")
        .and_then(Value::as_str)
        .ok_or_else(|| anyhow!("Workspace ID is required"))?;
    let mut status = client
        .request_value(
            "workspace.bufferGuard.acquire",
            &json!({"id": workspace_id, "operation": operation, "expectedInstanceId": payload.get("expectedInstanceId")}),
        )
        .await?;
    let id = status
        .get("guardId")
        .and_then(Value::as_str)
        .ok_or_else(|| anyhow!("Runtime did not return a buffer guard"))?
        .to_string();
    let result = async {
        if let Some(expected) = payload.get("expectedInstanceId").and_then(Value::as_str) {
            if status.get("workspaceInstanceId").and_then(Value::as_str) != Some(expected) {
                bail!("The runtime did not verify the expected workspace instance. Update the owner runtime or refresh the task before retrying.");
            }
        }
        tokio::time::timeout(std::time::Duration::from_secs(20), async {
            while status.get("ready") != Some(&Value::Bool(true)) {
                let blockers = status.get("blockers").and_then(Value::as_array).ok_or_else(|| anyhow!("Runtime did not verify editor buffers"))?;
                if !blockers.is_empty() {
                    bail!("Resolve editor buffers on the connected clients before continuing: {}", blockers.iter().map(|blocker| format!("{}: {}", blocker["path"].as_str().unwrap_or("editor"), blocker["reason"].as_str().unwrap_or("verification failed"))).collect::<Vec<_>>().join("; "));
                }
                if status.get("disconnectedClients").and_then(Value::as_u64) != Some(0) { bail!("A client disconnected during buffer verification. Prepare the operation again."); }
                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
                status = client.request_value("workspace.bufferGuard.status", &json!({"guardId": id})).await?;
            }
            Ok::<_, anyhow::Error>(())
        }).await.map_err(|_| anyhow!("Some connected editors did not respond. The workspace was preserved."))??;
        let mut request = payload.clone();
        request["bufferGuardId"] = json!(id);
        client.request_value(&format!("workspace.{operation}"), &request).await
    }.await;
    // A started mutation keeps its guard until completion even if the RPC
    // timed out. An unclaimed preparation can safely release or expire.
    let _ = client
        .request_value("workspace.bufferGuard.release", &json!({"guardId": id}))
        .await;
    result
}
